File: 0198 House Robber/leet_recursion_memo.py
class Solution(object):
    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        # top down, bottom up
        # tabulation, recursion, hybrid

        memo = {}

        def dp(i):
            if i == 0:
                return nums[0]
            elif i == 1:
                return max(nums[0], nums[1])
            else:
                if i in memo:
                    return memo[i]
                else:
                    p = max(dp(i - 1), dp(i - 2) + nums[i])
                    memo[i] = p
                    return p

        return dp(len(nums) - 1)

File: 0198 House Robber/leet_recursion_memo.go
func rob(nums []int) int {
    memo := make(map[int]int)

    var dp func(int)int

    dp = func(i int) int {
        if i == 0 {
            return nums[0]
        } else if i == 1 {
            return slices.Max(nums[:2])
        } else {
            if _, ok := memo[i]; ok == false {
                p := slices.Max([]int{dp(i-1), dp(i-2)+nums[i]})
                memo[i] = p
            }
            return memo[i]
        }
    }
    return dp(len(nums)-1)
}